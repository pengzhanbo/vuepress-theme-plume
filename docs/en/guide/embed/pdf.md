---
title: PDF Reader
icon: teenyicons:pdf-outline
createTime: 2025/10/08 13:30:53
permalink: /en/guide/embed/pdf/
---

## Overview

The theme supports embedding PDF files in markdown, enabling direct PDF reading within the page.

This feature is powered by [vuepress-plugin-md-power](../../config/plugins/markdown-power.md).

## Configuration

This feature is disabled by default. You need to enable it in the theme configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      pdf: true, // [!code ++]
    },
  })
})
```

## Syntax

The simplest syntax is as follows:

```md
@[pdf](url)
```

To open a specific page, add a page number after `pdf`:

```md
@[pdf 2](url)
```

Additional options can be added to `@[pdf ]` for more flexible control:

```md
@[pdf 2 no-toolbar width="100%" height="400px" ratio="16:9" zoom="100"](url)
```

- `no-toolbar` - Hide the toolbar
- `width` - Width, defaults to 100%
- `height` - Height, defaults to `auto`
- `ratio` - Aspect ratio, defaults to `16:9`, only effective when height is not specified
- `zoom` - Zoom level, percentage

## Examples

### Default

Input:

```md
@[pdf](https://plume.pengzhanbo.cn/files/sample.pdf)
```

Output:

@[pdf](/files/sample.pdf)

### Set Page Number to 2

Input:

```md
@[pdf 2](https://plume.pengzhanbo.cn/files/sample.pdf)
```

Output:

@[pdf 2 zoom="95"](/files/sample.pdf)

### Hide Toolbar

Input:

```md
@[pdf no-toolbar](https://plume.pengzhanbo.cn/files/sample.pdf)
```

Output:

@[pdf no-toolbar](/files/sample.pdf)

### 90% Zoom Level

Input:

```md
@[pdf zoom="90"](https://plume.pengzhanbo.cn/files/sample.pdf)
```

Output:

@[pdf zoom="90"](/files/sample.pdf)

### 21:29 Aspect Ratio

Input:

```md
@[pdf zoom="95" ratio="21:29"](https://plume.pengzhanbo.cn/files/sample.pdf)
```

Output:

@[pdf zoom="95" ratio="21:29"](/files/sample.pdf)
