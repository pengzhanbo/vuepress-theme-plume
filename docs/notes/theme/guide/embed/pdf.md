---
title: PDF 阅读
icon: teenyicons:pdf-outline
createTime: 2024/03/28 13:30:53
permalink: /guide/embed/pdf/
---

## 概述

主题支持在 markdown 中嵌入 PDF 文件，它能够在页面中直接阅读 PDF 。

该功能由 [vuepress-plugin-md-power](../../config/plugins/markdown-power.md) 提供支持。

## 配置

该功能默认不启用。你需要在主题配置中开启。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      pdf: true, // [!code ++]
    },
  })
})
```

## 语法

最简单的语法如下：

```md
@[pdf](url)
```

当需要打开特定页面时，在 `pdf` 后面跟随一个 页数。

```md
@[pdf 2](url)
```

还可以添加更多的 选项到 `@[pdf ]` 中，更灵活的控制行为。

```md
@[pdf 2 no-toolbar width="100%" height="400px" ratio="16:9" zoom="100"](url)
```

- `no-toolbar` - 不显示工具栏
- `width` - 宽度，默认为 100%
- `height` - 高度，默认为 `auto`
- `ratio` - 宽高比, 默认为 `16:9`, 仅当未指定高度时生效
- `zoom` - 缩放比例, 百分比。

## 示例

### 默认

输入：

```md
@[pdf](https://plume.pengzhanbo.cn/files/sample.pdf)
```

输出：

@[pdf](/files/sample.pdf)

### 设置页码为 2

输入：

```md
@[pdf 2](https://plume.pengzhanbo.cn/files/sample.pdf)
```

输出：

@[pdf 2 zoom="95"](/files/sample.pdf)

### 不显示工具栏

输入：

```md
@[pdf no-toolbar](https://plume.pengzhanbo.cn/files/sample.pdf)
```

输出：

@[pdf no-toolbar](/files/sample.pdf)

### 缩放比 90%

输入：

```md
@[pdf zoom="90"](https://plume.pengzhanbo.cn/files/sample.pdf)
```

输出：

@[pdf zoom="90"](/files/sample.pdf)

### 宽高比 21:29

输入：

```md
@[pdf zoom="95" ratio="21:29"](https://plume.pengzhanbo.cn/files/sample.pdf)
```

输出：

@[pdf zoom="95" ratio="21:29"](/files/sample.pdf)
