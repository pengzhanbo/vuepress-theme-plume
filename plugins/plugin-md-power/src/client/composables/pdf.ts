/**
 * Fork for https://github.com/vuepress-theme-hope/vuepress-theme-hope/blob/main/packages/components/src/client/utils/viewPDF.ts
 *
 * The MIT License (MIT)
 * Copyright (C) 2021 - PRESENT by Mr.Hope<mister-hope@outlook.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import type { PDFEmbedType, PDFTokenMeta } from '../../shared/index.js'
import { checkIsiPad, checkIsMobile, checkIsSafari } from '@vuepress/helper/client'
import { withBase } from 'vuepress/client'
import { ensureEndingSlash, isLinkHttp } from 'vuepress/shared'
import { pluginOptions } from '../options.js'

function queryStringify(options: PDFTokenMeta): string {
  const { page, noToolbar, zoom } = options
  const params = [
    `page=${page}`,
    `toolbar=${noToolbar ? 0 : 1}`,
    `zoom=${zoom}`,
  ]

  let queryString = params.join('&')
  if (queryString)
    queryString = `#${queryString}`

  return queryString
}

export function renderPDF(
  el: HTMLElement,
  url: string,
  embedType: PDFEmbedType,
  options: PDFTokenMeta,
): void {
  if (!pluginOptions.pdf)
    return
  url = isLinkHttp(url)
    ? url
    : new URL(withBase(url), typeof location !== 'undefined' ? location.href : '').toString()

  const pdfOptions = pluginOptions.pdf === true ? {} : pluginOptions.pdf
  pdfOptions.pdfjsUrl ??= 'https://static.pengzhanbo.cn/pdfjs/'
  const pdfjsUrl = `${ensureEndingSlash(withBase(pdfOptions.pdfjsUrl))}web/viewer.html`

  const queryString = queryStringify(options)

  const source = embedType === 'pdfjs'
    ? `${pdfjsUrl}?file=${url}${queryString}`
    : `${url}${queryString}`

  const tagName = embedType === 'pdfjs' || embedType === 'iframe'
    ? 'iframe'
    : 'embed'

  el.innerHTML = ''
  const pdf = document.createElement(tagName)

  pdf.className = 'pdf-viewer';
  (pdf as any).type = 'application/pdf'
  pdf.title = options.title || 'PDF Viewer'
  pdf.src = source

  if (pdf instanceof HTMLIFrameElement)
    pdf.allow = 'fullscreen'

  el.appendChild(pdf)
}

export function usePDF(
  el: HTMLElement,
  url: string,
  options: PDFTokenMeta,
): void {
  if (typeof window === 'undefined' || !window?.navigator?.userAgent)
    return

  const { navigator } = window
  const { userAgent } = navigator

  const isModernBrowser = typeof window.Promise === 'function'

  // Quick test for mobile devices.
  const isMobileDevice = checkIsiPad(userAgent) || checkIsMobile(userAgent)

  // Safari desktop requires special handling
  const isSafariDesktop = !isMobileDevice && checkIsSafari(userAgent)

  const isFirefoxWithPDFJS
    = !isMobileDevice
      && /firefox/iu.test(userAgent)
      && userAgent.split('rv:').length > 1
      ? Number.parseInt(userAgent.split('rv:')[1].split('.')[0], 10) > 18
      : false

  // Determines whether PDF support is available
  const supportsPDFs
    // As of Sept 2020 no mobile browsers properly support PDF embeds
    = !isMobileDevice
    // We're moving into the age of MIME-less browsers. They mostly all support PDF rendering without plugins.
      && (isModernBrowser
      // Modern versions of Firefox come bundled with PDFJS
        || isFirefoxWithPDFJS)

  if (!url)
    return

  if (supportsPDFs || !isMobileDevice) {
    const embedType = isSafariDesktop ? 'iframe' : 'embed'
    return renderPDF(el, url, embedType, options)
  }

  return renderPDF(el, url, 'pdfjs', options)
}
