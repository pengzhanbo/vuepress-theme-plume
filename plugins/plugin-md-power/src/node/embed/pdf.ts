/**
 * @[pdf](/xxx)
 * @[pdf 1](/xxx)
 * @[pdf 1 no-toolbar width="100%" height="600px" zoom="1" ratio="1:1"](/xxx)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { PDFTokenMeta } from '../../shared/index.js'
import { path } from 'vuepress/utils'
import { parseRect } from '../utils/parseRect.js'
import { resolveAttrs } from '../utils/resolveAttrs.js'
import { createEmbedRuleBlock } from './createEmbedRuleBlock.js'

export const pdfPlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<PDFTokenMeta>(md, {
    type: 'pdf',
    // eslint-disable-next-line regexp/no-super-linear-backtracking
    syntaxPattern: /^@\[pdf(?:\s+(\d+))?([^\]]*)\]\(([^)]*)\)/,
    meta([, page, info = '', src = '']) {
      const { attrs } = resolveAttrs(info)
      return {
        src,
        page: +page || 1,
        noToolbar: Boolean(attrs.noToolbar ?? false),
        zoom: +attrs.zoom || 50,
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '',
        ratio: attrs.ratio ? parseRect(attrs.ratio) : '',
        title: path.basename(src || ''),
      }
    },
    content({ title, src, page, noToolbar, width, height, ratio, zoom }) {
      return `<PDFViewer src="${src}" title="${title}" :page="${page}" :no-toolbar="${noToolbar}" width="${width}" height="${height}" ratio="${ratio}" :zoom="${zoom}" />`
    },
  })
}
