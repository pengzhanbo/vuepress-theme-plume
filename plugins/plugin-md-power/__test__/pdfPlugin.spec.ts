import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { pdfPlugin } from '../src/node/embed/pdf.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(pdfPlugin)
}

describe('pdfPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[pdf](foo.pdf)

@[pdf no-toolbar](foo.pdf)

@[pdf 2](foo.pdf)

@[pdf 2 no-toolbar](foo.pdf)

@[pdf 2 no-toolbar width="100%" height="600px" zoom="1"](foo.pdf)

@[pdf 2 no-toolbar width="100%" zoom="1" ratio="1:1"](foo.pdf)
`

    expect(md.render('@[pdf](foo.pdf)')).toMatchSnapshot()
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[pdf]()

@[pdf]xxx

@[ pdf]()

@[pdf]
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
