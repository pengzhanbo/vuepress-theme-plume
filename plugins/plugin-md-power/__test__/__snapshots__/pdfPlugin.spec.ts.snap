// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pdfPlugin > should not work 1`] = `
"<PDFViewer src="" title="" :page="1" :no-toolbar="false" width="100%" height="" ratio="" :zoom="50" /><p>@[pdf]xxx</p>
<p>@<a href=""> pdf</a></p>
<p>@[pdf]</p>
"
`;

exports[`pdfPlugin > should work 1`] = `"<PDFViewer src="foo.pdf" title="foo.pdf" :page="1" :no-toolbar="false" width="100%" height="" ratio="" :zoom="50" />"`;

exports[`pdfPlugin > should work 2`] = `"<PDFViewer src="foo.pdf" title="foo.pdf" :page="1" :no-toolbar="false" width="100%" height="" ratio="" :zoom="50" /><PDFViewer src="foo.pdf" title="foo.pdf" :page="1" :no-toolbar="true" width="100%" height="" ratio="" :zoom="50" /><PDFViewer src="foo.pdf" title="foo.pdf" :page="2" :no-toolbar="false" width="100%" height="" ratio="" :zoom="50" /><PDFViewer src="foo.pdf" title="foo.pdf" :page="2" :no-toolbar="true" width="100%" height="" ratio="" :zoom="50" /><PDFViewer src="foo.pdf" title="foo.pdf" :page="2" :no-toolbar="true" width="100%" height="600px" ratio="" :zoom="1" /><PDFViewer src="foo.pdf" title="foo.pdf" :page="2" :no-toolbar="true" width="100%" height="" ratio="1:1" :zoom="1" />"`;
